peps/pep-8106.rst

```diff
@@ -0,0 +1,171 @@
+that allows whitespace, e.g. to be lenient on user input in a web form::
+
+    from typing import Annotated
+
+    from pydantic import BaseModel, parse
+
+    class Item(BaseModel):
+        price: Annotated[float, parse(str).transform(str.strip).transform(float)]
+
+    Model.model_validate_json('{"value":"1.0 "}')  # okay!
+
+Staying open to new future use cases
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This proposal is designed to be open to future use cases.
+By adding a re-usable generic that is not constrained to specific use cases
+we can allow for future use cases that we cannot predict today, just like :pep:`593` did originally.
+
+This is in contrast to proposals addressing more narrow use cases like :pep:`712` which introduced
+a new ``converter`` parameter to ``dataclasses.field``.
+One can imagine that if in the future we arrive at a de-facto standard for a
+converter that uses ``AnnotatedMeta`` and we can agree upon a way to express allowing an input
+type that is different than the field type we could then standardize that via it's own PEP,
+but this PEP does not attempt to do so so that we can let these details be worked out more organically.
+
+Specification
+-------------
+
+Implementation
+~~~~~~~~~~~~~~
+
+The following class is added to the typing module::
+
+    class AnnotatedMeta[FieldT]:
+        ...
+
+Libraries and tools that want to support ``Annotated`` type checking have their
+metadata inherit from it::
+
+    from typing import AnnotatedMeta, Any, Callable, overload
+
+    class Int64(AnnotatedMeta[int]):
+        ...
+
+    class Validate[In: Any, Out](AnnotatedMeta[Out]):
+        @overload
+        def parse[NewOut](self, tp: type[NewOut]) -> Validate[In, NewOut]: ...
+
+        @overload
+        def parse[NewOut](self: Validate[In, NewOut]) -> Validate[In, NewOut]: ...
+
+        def transform[NewOut](self, func: Callable[[Out], NewOut]) -> Validate[In, NewOut]: ...
+    
+    parse = Validate().parse  # note that Out is unsubstituted
+
+
+    # okay
+    Annotated[int, Int64()]
+
+    # error
+    Annotated[float, Int64()]
+
+    # okay, Validate[Any, str] is explicitly created and str.lower(str) is valid
+    Annotated[str, parse(str).transform(str.lower)]
+
+    # error, Validate[Any, str] is explicitly created and str / 1 is invalid
+    Annotated[str, parse(str).transform(lambda x: x / 1)]
+
+    # okay, Validate[Any, float] is inferred because type parameter is not substituted
+    Annotated[float, parse().transform(lambda x: x / 1)]
+
+    # error, float is inferred and str.lower does not work on float
+    Annotated[float, parse().transform(str.lower)]
+
+
+Type checking behavior
+~~~~~~~~~~~~~~~~~~~~~~
+
+If type checkers support type checking the metadata in ``Annotated`` they should
+follow the behavior that if ``AnnotatedMeta`` is un-parameterized it's type parameter
+should take the value of the type in ``Annotated``.
```

Comment:
I don't understand what this sentence means. I think the spec should be something like this: When a type checker sees an annotation of the form `Annotated[T, metadata]`, where `metadata` can stand for any number of metadata objects, then for each element in the metadata, if it is an instance of `AnnotatedMeta`, it should check that `T` is consistent with the type parameter of `AnnotatedMeta`.

Reply:
I'm hoping we can achieve something like:

```python
Annotated[str, Validate().transform(str.lower()]
#                       ↑ because we didn't parametrize Validate (e.g. Validate[int]) this is Validate[str]
```

Reply:
Anything after `transform(str.lower)` would be recognized as as a string but the call inside of `transform` would be untyped. That is, if I do `Validate().transform(lambda x: x.lower())` the best I can do is make `x` `Any` so you don't get false positive type errors but if would be nicer if it auto-completed `.lower()` and type checked properly.